src/rgw/driver/rados/rgw_sync.cc

```diff
@@ -1587,15 +1590,68 @@ class RGWMetaSyncShardCR : public RGWCoroutine {
       /* sync! */
       do {
         if (!lease_cr->is_locked()) {
-          tn->log(1, "lease is lost, abort");
-          lost_lock = true;
-          break;
+          if (sync_env->bid_manager->is_highest_bidder(shard_id, ceph::coarse_mono_clock::now())) {
```

Comment:
the `lease_cr->is_locked()` check is testing whether the `ContinuousLeaseCR` has timed out or failed. i don't think we should change the behavior in that case; it should just break out of this do-while loop. below, the `lost_lock` flag will prevent us from finishing full sync and switching to incremental

this check for `!is_highest_bidder()` should act similarly: we should break out of the loop, but still call `marker_tracker->flush()` to update the full sync status after all the sync coroutines finish

we might add a separate `bool lost_bid` variable for this, and insert that new logic after the loop:
```diff
       while (num_spawned() > 1) {
         yield wait_for_child();
         collect_children();
       }
 
-      if (!lost_lock) {
+      if (lost_bid) {
+        yield call(marker_tracker->flush());
+      } else if (!lost_lock) {
         /* update marker to reflect we're done with full sync */
```
and further down:
```diff
-      if (lost_lock) {
+      if (lost_lock || lost_bid) {
         return -EBUSY;
       }
 
       tn->log(10, "full sync complete");
```
the same should apply for `incremental_sync()` below

does this make sense?

Reply:
@cbodley understood the `lost_lock` part. As for calling RGWContinouseLeaseCR(), do you mean we don't want to be nesting 'is_highest_bidder()' logic under the '!lease_cr->is_locked()' condition? That would mean we are checking if we are outbid or not while we still have the lease?

Reply:
> That would mean we are checking if we are outbid or not while we still have the lease?

that's right. once sync acquire this lock, `RGWContinouseLeaseCR` will keep trying to renew it indefinitely. but if we hear from another gateway with a higher bid, we should wind down and unlock so that other gateway can take over

the `lost_lock` case should only happen in extreme cases, where our lease times out because the osd takes longer than 1 minute (or rgw_sync_lease_period/2) to respond to our lock request

Reply:
got it. thanks!

Reply:
may be we should call `marker_tracker->flush()` even in the lost_lock case?

Reply:
not sure; i think data sync does that, but there are tradeoffs. maybe we should discuss that separately?